hack/make-release.sh

```diff
@@ -34,6 +34,10 @@ if [[ "$TAG" != v* ]]; then
     fail "\$TAG must start with 'v', e.g. v0.1.0 (got: $TAG)"
 fi
 
+# make sure local source is up to date
+git checkout main
+git pull
```

Comment:
Instead of silently changing branch (if applicable) and pulling (if applicable), may I suggest we loudly fail instead?

e.g. for the branch
```
if [[ <branch is not main> ]]; then
    echo "error: need to be checked in on the main branch"
    exit 1
fi
```

Could also have `if [[ <branch not in sync with upstream> ]]; then` etc.

That way the "user" of the release script realize before continuing with the script that "ah, hold on, what is happening here, am I in the wrong place?" and have the time to double check some more stuff before proceeding with the rest of the releasing script.

Reply:
I see what you mean, but here the goal is to make sure it could be automated, without any user interaction. Ultimately, this script needs to be executed by a fully automated process, no local script execution. So this `git checkout` + `git pull` should mimic this part, don't you think?

Reply:
I see what you mean but picture this scenario:
- You want to release a new version of Online Boutique
- You were working on a new feature / bug fix / whatever on a separate branch
- You forgot to change branch (but you have a few outstanding files not tracked by git yet)
- You run the releasing script that does some "magic" by checking out the `main` branch
- Uh oh... The files you were working on have carried over to the `main` branch (well, in fact they're not in _any_ branch still they're not tracked)
- Boom. You just released Online Boutique with unintended features / bug fixes / whatever :)

On another hand, the script could continue doing magic by running say `git add *` and `git stash` before the `git checkout` but still, I think that might create a lot of confusion (from the "user" which had no idea they were on the wrong branch).

I'll take an extra 10 seconds to manually change branch vs. potentially releasing something unintended and/or losing "in progress" files from an unrelated feature.

At minimum, there should be a hard-fail if there are _any_ uncommited changes:
```
# ensure there are no uncommitted changes
if [[ $(git status -s | wc -l) -gt 0 ]]; then
    echo "error: can't have uncommitted changes"
    exit 1
fi
```

Reply:
Good point, just added this check to make sure about that. Thansk!